keps/sig-aws/0014-iam-credentials-provider.md

```diff
@@ -0,0 +1,209 @@
+
+Pods will be associated to IAM roles using a Custom Resource: an IAMBinding. This will map between a Pod’s service account and the IAM role.
+
+**Restricting roles to namespaces**
+
+Our proposal would allow IAM assume policy to be set similarly to below for controlling which service accounts can assume which roles:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Action": "sts:AssumeRole",
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": "arn:aws:iam::123456789012:role/kiam_role"
+      },
+      "Condition": {
+        "StringLike": {
+          "aws:userid": "cluster.namespace.service-account-name"
+        }
+      }
+    }
+  ]
+}
+```
+
+#### Proxy
+Any Pod needing credentials will have an init container and sidecar. 
+
+The init container will ensure that other containers in the Pod when accessing `http://169.254.170.2` will have requests forwarded to the sidecar. The init container can also request initial credentials ensuring that the pod is entitled to use credentials for a role configured through the IAMBinding.
+
+Once the init container has obtained the initial credentials the sidecar can run and refresh credentials periodically. 
+
+The Proxy will read the Pod’s service account token and pass it to the Server for a TokenReview allowing the Server to attest to the Pod identity that is executing the Proxy against the Kubernetes API Server.
+
+#### Admission Controller
+This is responsible for identifying whether a Pod matches an IAMBinding and, if so, injects the init and sidecar.
+
+#### Intercepting AWS APIs
+We intend to follow the ECS credential provider, which uses an endpoint on http://169.254.170.2  that is typically provided by the ECS agent. In some of the standard SDKs this adds additional retry behaviour rather than the hard fail if the instance metadata API fails. Our goal would be to set environment variables that would trigger this behaviour and so mimic IAM integration provided in Amazon ECS.
+
+There is potential for the workload container to start faster than the sidecar leading to failed requests. This is why we preferred use of the ECS credential provider, as it will retry on failure. The [SIG-Apps Sidecar proposal](https://github.com/kubernetes/community/pull/2148) would also help such an issue.
+
+Using a per-Pod proxy helps mitigate many of the shortcomings of the existing DaemonSet-style solutions: IAM permissions can be checked before pods start, updates to the proxy are associated with the lifecycle of the Pod process, stronger identity attestation and removing inconsistency issues from tracking pods against the watcher. Existing solutions also often rely on per-node agents which, when restarted, can cause problems for the Pods running on the same nodes. Running a proxy sidecar against the Pod mitigates this.
+
+Our assumption is that proxies can be injected in the same way Istio injects it’s proxy (iptables from the init container).
```

Comment:
does this require high permission levels for the init container?

Reply:
it would be good to get feedback from the istio team on the viability of this approach, challenges they've hit in different environments where initContainers are not allowed those privilege levels, interactions with different types of network environments, etc

Reply:
As above, it does need higher permissions- I think `NET_ADMIN` would suffice. But, it would only need the init container to run with additional privileges.

Agreed- be great to know whether this is a legitimate approach, or something we should/could accomplish in a different way.